tempesta_fw/str.h

```diff
@@ -98,40 +94,45 @@
  * @flags	- 3 most significant bytes for number of chunks of compound
  * 		  string and the least significant byte for flags;
  */
-typedef struct {
-	void		*ptr;
+typedef struct TfwStr {
 	struct sk_buff	*skb;
 	unsigned long	len;
 	unsigned char	eolen;
-	unsigned int	flags;
+
+	unsigned int	chunknum : 24;
+	unsigned int	flags : 8;
+	union {
+	char *data;
+	struct TfwStr *chunks;
+};
 } TfwStr;
 
-#define DEFINE_TFW_STR(name, val) TfwStr name = { (val), NULL,		\
-						  sizeof(val) - 1, 0 }
-#define TFW_STR_FROM(s)         ((TfwStr){(char*)s, NULL, strlen(s)})
+#define DEFINE_TFW_STR(name, val) TfwStr name = { .data = (val), .skb = NULL\
+, .len = sizeof(val) - 1, .chunknum = 0, .flags = 0 }
+#define TFW_STR_FROM(s)         ((TfwStr){ .data = (char *)s, .skb = NULL,\
+  .len = sizeof(s) - 1})
 
 /* Use this with "%.*s" in printing calls. */
-#define PR_TFW_STR(s)		(int)min(20UL, (s)->len), (char *)(s)->ptr
+#define PR_TFW_STR(s)		(int)min(20UL, (s)->len), (s)->data
 
 /* Numner of chunks in @s. */
-#define TFW_STR_CHUNKN(s)	((s)->flags >> TFW_STR_CN_SHIFT)
+#define TFW_STR_CHUNKN(s)	((s)->chunknum)
 #define TFW_STR_CHUNKN_LIM(s)	((s)->flags >= __TFW_STR_CN_MAX)
```

Comment:
Are you sure?  How the flags are related to the upper limit on the number of chunks in a `TfwStr{}` string?